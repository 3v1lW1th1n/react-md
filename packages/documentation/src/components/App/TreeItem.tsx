import * as React from "react";
import * as ReactDOM from "react-dom";
import { DefaultTreeItemRenderer, TreeViewData, IDefaultTreeItemRendererProps } from "@react-md/tree-view";
import { MagicTooltip } from "@react-md/tooltip";

export interface ITreeItemState {
  tooltipped: boolean;
}

export default class TreeItem extends React.Component<IDefaultTreeItemRendererProps, ITreeItemState> {
  private item: HTMLElement | null;
  constructor(props: IDefaultTreeItemRendererProps) {
    super(props);

    this.item = null;
    this.state = { tooltipped: false };
  }

  public componentDidMount() {
    this.item = ReactDOM.findDOMNode(this) as HTMLElement;
    this.scrollIntoView();

    const tooltipped = this.checkTooltip(this.props);
    if (this.state.tooltipped !== tooltipped) {
      this.setState({ tooltipped });
    }
  }

  public componentDidUpdate(prevProps: IDefaultTreeItemRendererProps, prevState: ITreeItemState) {
    const tooltipped = this.checkTooltip(this.props);
    if (this.state.tooltipped !== tooltipped) {
      this.setState({ tooltipped });
    }
  }

  public render() {
    const { tooltipped } = this.state;
    const { children, ...props } = this.props;

    let tooltip;
    let tooltipId;
    if (tooltipped) {
      let { depth, itemIndex } = props;
      depth += 1;
      itemIndex += 1;
      tooltipId = `navigation-item-${depth}-${itemIndex}-tooltip`;
      tooltip = <MagicTooltip id={tooltipId}>{children}</MagicTooltip>;
    }

    return (
      <DefaultTreeItemRenderer {...props} aria-describedby={tooltipId}>
        {children}
        {tooltip}
      </DefaultTreeItemRenderer>
    );
  }

  private checkTooltip = ({ depth, itemIndex }: IDefaultTreeItemRendererProps) => {
    if (!this.item) {
      return false;
    }

    const node = this.item.querySelector(".rmd-list-item__text") as HTMLElement;
    return !!node && node.scrollWidth > node.offsetWidth;
  };

  private scrollIntoView = () => {
    if (!this.item || !this.props.selected) {
      return;
    }

    window.requestAnimationFrame(() => {
      const navigation = document.getElementById("navigation") as HTMLElement;
      const header = navigation && navigation.querySelector("header") as HTMLElement;
      if (navigation && this.item && header) {
        navigation.scrollTop = this.item.offsetTop - header.offsetHeight;
      }
    });
  };
}
