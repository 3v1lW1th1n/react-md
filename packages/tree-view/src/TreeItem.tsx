import * as React from "react";
import cn from "classnames";
import { StatesConsumer } from "@react-md/states";

/**
 * The `TreeItem` component is an extremely simple component that just renders an `li` element
 * with the "base" tree item props and any children provided. This should be used in combination
 * with the `TreeItemContent` and `TreeGroup` components to get a fully functional tree item.
 *
 * If you want to render the treeitem as a link, please use the `TreeLinkItem` component instead
 * of this one.
 */
export interface ITreeItemBaseProps {
  /**
   * An optional aria-expanded attribute to apply to the tree item. This should only be provided as the value "true"
   * and only if it is currently expanded. It should be `undefined` otherwise.
   *
   * @docgen
   */
  "aria-expanded"?: "true";

  /**
   * The current level (depth) for the tree item.
   *
   * @docgen
   */
  "aria-level": number;

  /**
   * The tree item's current position within the parent treeview or a sub-group. This should be a number starting
   * from `1`.
   *
   * @docgen
   */
  "aria-posinset": number;

  /**
   * The size of the treeview or sub-group that the tree item is in. This should be a number starting from `1`.
   *
   * @docgen
   */
  "aria-setsize": number;

  /**
   * The tabindex for the tree item. When working with a single-selection tree view, **only 1 treeitem** can have a
   * tab index of `0` while all other treeitems should have a tab index of `-1`.
   *
   * It is generally recommended to keep this prop `undefined` and let the `selected` prop handle setting the correct
   * `tabIndex` instead. However, you can manually override the built-in behavior by setting this to valid number.
   *
   * @docgen
   */
  tabIndex?: 0 | -1;

  /**
   * Boolean if the tree item is current selected. This will update the styles to be have the selected state as well
   * as changing the `tabIndex` from `-1` to `0` so it is focusable. If the `tabIndex` prop is defined, that value
   * will **always** be used instead.
   *
   * @docgen
   */
  selected?: boolean;
}

export type ITreeItemProps = React.HTMLAttributes<HTMLLIElement> & ITreeItemBaseProps;

export interface ITreeItemDefaultProps {
  selected: boolean;
}

export type TreeItemWithDefaultProps = ITreeItemProps & ITreeItemDefaultProps;

const TreeItem: React.SFC<ITreeItemProps> = providedProps => {
  const {
    className,
    selected,
    tabIndex: propTabIndex,
    children,
    onMouseDown,
    onMouseUp,
    onTouchStart,
    onTouchEnd,
    onKeyDown,
    onKeyUp,
    ...props
  } = providedProps as TreeItemWithDefaultProps;

  let tabIndex = propTabIndex;
  if (typeof tabIndex !== "number") {
    tabIndex = selected ? 0 : -1;
  }

  return (
    <StatesConsumer
      selected={selected}
      className={cn("rmd-tree-item", className)}
      onMouseDown={onMouseDown}
      onMouseUp={onMouseUp}
      onTouchStart={onTouchStart}
      onTouchEnd={onTouchEnd}
      onKeyUp={onKeyUp}
      onKeyDown={onKeyDown}
    >
      {statesProps => (
        <li {...statesProps} {...props} role="treeitem" tabIndex={tabIndex}>
          {children}
        </li>
      )}
    </StatesConsumer>
  );
};

export default TreeItem;
