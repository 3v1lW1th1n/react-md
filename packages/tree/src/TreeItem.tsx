import React, { FC, forwardRef, useCallback } from "react";
import cn from "classnames";
import { FontIcon } from "@react-md/icon";
import { ListItemChildren, ListElement } from "@react-md/list";
import { useInteractionStates } from "@react-md/states";
import { bem, WithForwardedRef } from "@react-md/utils";

import TreeGroup from "./TreeGroup";
import TreeItemExpanderIcon from "./TreeItemExpanderIcon";
import { TreeItemProps } from "./types";

type WithRef = WithForwardedRef<HTMLLIElement>;
type DefaultProps = Required<
  Pick<
    TreeItemProps,
    | "contentComponent"
    | "expanderLeft"
    | "expanderIcon"
    | "disabled"
    | "height"
    | "threeLines"
    | "textChildren"
  >
>;
type WithDefaultProps = TreeItemProps & DefaultProps & WithRef;

const block = bem("rmd-tree-item");

/**
 * This component renders an item within a tree with optional child items. This should almost always
 * be used from the `itemRenderer` prop from the `Tree` component as it provides a lot of the required
 * a11y props for you.
 */
const TreeItem: FC<TreeItemProps & WithRef> = providedProps => {
  const {
    id,
    className: propClassName,
    liStyle,
    liClassName,
    liRef,
    forwardedRef,
    depth,
    listSize,
    itemIndex,
    renderChildItems,
    expanded,
    selected,
    focused,
    expanderIcon,
    expanderLeft,
    textClassName,
    secondaryTextClassName,
    textChildren,
    primaryText,
    secondaryText,
    leftIcon: propLeftIcon,
    leftAvatar,
    rightIcon: propRightIcon,
    rightAvatar,
    forceIconWrap,
    height,
    threeLines,
    children,
    contentComponent: Content,
    isLink: propIsLink,
    disabled,
    readOnly,
    onFocus,
    ...props
  } = providedProps as WithDefaultProps;

  const isLink =
    typeof propIsLink === "boolean"
      ? propIsLink
      : typeof Content !== "string" || Content === "a";

  const { ripples, className, handlers } = useInteractionStates({
    disabled,
    className: propClassName,
    handlers: isLink ? props : undefined,
    disableSpacebarClick: isLink,
  });

  let group;
  let leftIcon = propLeftIcon;
  let rightIcon = propRightIcon;
  if (renderChildItems) {
    const icon = (
      <TreeItemExpanderIcon rotated={expanded}>
        {expanderIcon}
      </TreeItemExpanderIcon>
    );
    if (expanderLeft) {
      leftIcon = icon;
    } else {
      rightIcon = icon;
    }

    group = <TreeGroup collapsed={!expanded}>{renderChildItems()}</TreeGroup>;
  }

  const handleFocus = useCallback(
    (event: React.FocusEvent<HTMLLIElement>) => {
      if (onFocus) {
        onFocus(event);
      }

      event.preventDefault();
      const tree = event.currentTarget.closest('[role="tree"]');
      if (tree) {
        (tree as ListElement).focus();
      }
    },
    [onFocus]
  );

  const a11y = {
    "aria-expanded": renderChildItems ? expanded : undefined,
    "aria-level": depth + 1,
    "aria-setsize": listSize,
    "aria-posinset": itemIndex + 1,
    "aria-disabled": disabled ? "true" : undefined,
    id,
    role: "treeitem",
    tabIndex: -1,
    ...handlers,
    onFocus: handleFocus,
  };
  const noA11y = { role: "none" };

  return (
    <li
      {...(isLink ? noA11y : a11y)}
      ref={liRef}
      style={liStyle}
      className={cn(block(), liClassName)}
    >
      <Content
        {...props}
        {...(isLink ? a11y : undefined)}
        ref={forwardedRef}
        className={cn(
          block("content", {
            link: isLink,
            clickable: !disabled && !readOnly,
            [height]: height !== "auto" && height !== "normal",
            "three-lines": !!secondaryText && threeLines,
            selected,
            focused,
          }),
          className
        )}
      >
        <ListItemChildren
          textClassName={textClassName}
          secondaryTextClassName={secondaryTextClassName}
          textChildren={textChildren}
          primaryText={primaryText}
          secondaryText={secondaryText}
          leftIcon={leftIcon}
          leftAvatar={leftAvatar}
          rightIcon={rightIcon}
          rightAvatar={rightAvatar}
          forceIconWrap={forceIconWrap}
        >
          {children}
        </ListItemChildren>
        {ripples}
      </Content>
      {group}
    </li>
  );
};

const defaultProps: DefaultProps = {
  contentComponent: "span",
  expanderLeft: false,
  expanderIcon: <FontIcon>keyboard_arrow_down</FontIcon>,
  disabled: false,
  height: "auto",
  threeLines: false,
  textChildren: true,
};

TreeItem.defaultProps = defaultProps;

if (process.env.NODE_ENV !== "production") {
  TreeItem.displayName = "TreeItem";

  let PropTypes;
  try {
    PropTypes = require("prop-types");
  } catch (e) {}

  if (PropTypes) {
    TreeItem.propTypes = {
      id: PropTypes.string.isRequired,
      depth: PropTypes.number.isRequired,
      itemIndex: PropTypes.number.isRequired,
      listSize: PropTypes.number.isRequired,
      selected: PropTypes.bool.isRequired,
      expanded: PropTypes.bool.isRequired,
      focused: PropTypes.bool.isRequired,
      renderChildItems: PropTypes.func,
      liStyle: PropTypes.object,
      liClassName: PropTypes.string,
      isLink: PropTypes.bool,
      contentComponent: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.func,
        PropTypes.object,
      ]),
      disabled: PropTypes.bool,
      expanderLeft: PropTypes.bool,
      expanderIcon: PropTypes.node,
      height: PropTypes.oneOf([
        "auto",
        "normal",
        "medium",
        "large",
        "extra-large",
      ]),
      threeLines: PropTypes.bool,
      textChildren: PropTypes.bool,
    };
  }
}

export default forwardRef<HTMLLIElement, TreeItemProps>((props, ref) => (
  <TreeItem {...props} forwardedRef={ref} />
));
