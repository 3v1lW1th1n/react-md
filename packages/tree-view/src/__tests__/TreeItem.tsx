import * as React from "react";
import renderer from "react-test-renderer";

import TreeItem from "../TreeItem";

const BASE_PROPS = {
  "aria-level": 1,
  "aria-posinset": 1,
  "aria-setsize": 1,
};

// TODO: Write more tests once enzyme supports the new react context API.
describe("TreeItem", () => {
  it("should render correctly with the provided props", () => {
    expect(renderer.create(<TreeItem {...BASE_PROPS}>content</TreeItem>).toJSON()).toMatchSnapshot();
  });
});
