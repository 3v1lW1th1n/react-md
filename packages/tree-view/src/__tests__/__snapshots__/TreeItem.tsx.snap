// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeItem should render correctly with the provided props 1`] = `
<li
  aria-level={1}
  aria-posinset={1}
  aria-setsize={1}
  className="rmd-tree-item rmd-states--simple"
  disabled={false}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  role="treeitem"
  tabIndex={-1}
>
  content
</li>
`;
